huntsman/observatory.py

```diff
@@ -472,43 +355,217 @@ def _create_autoguider(self):
-        field = Field('Evening Flats', flat_coords.to_string('hmsdms'))
-        flat_obs = DitheredObservation(field, exptime=1. * u.second)
-        flat_obs.seq_time = utils.current_time(flatten=True)
+        return obs
 
-        if isinstance(flat_obs, DitheredObservation):
+    def _take_autoflats(self, cameras, observation, safety_func, min_counts=5000,
+                        max_counts=15000, bias=1000, max_exptime=60*u.second,
+                        target_scaling=0.5, max_num_exposures=10, min_exptime=1*u.second,
+                        *args, **kwargs):
+        """Take flat fields iteratively by automatically estimating exposure times.
 
-            dither_coords = dither.get_dither_positions(flat_obs.field.coord,
-                                                        n_positions=n_positions,
-                                                        pattern=dither.dice9,
-                                                        pattern_offset=dither_pattern_offset,
-                                                        random_offset=dither_random_offset)
+        Args:
+            cameras (dict): Dict of camera name: Camera pairs.
+            filter_names (dict): Dict of filter name for each camera.
+            safety_func (func): Boolean function that returns True only if safe to continue.
+        """
+        target_counts = min_counts + target_scaling * (max_counts - min_counts)
 
-            self.logger.debug("Dither Coords for Flat-field: {}".format(dither_coords))
+        # Setup containers with initial values
+        exptimes = {cam_name: [1. * u.second] for cam_name in cameras.keys()}
+        finished = {cam_name: False for cam_name in cameras.keys()}
+        n_good_exposures = {cam_name: 0 for cam_name in cameras.keys()}
 
-            fields = [Field('Dither{:02d}'.format(i), coord)
-                      for i, coord in enumerate(dither_coords)]
-            exptimes = [flat_obs.exptime for coord in dither_coords]
+        # Loop until conditions are met to finish flat-fielding
+        while not all(finished.values()):
+
+            if not safety_func():
+                self.logger.info('Stopping flat-fielding as no longer safe.')
+                break
 
-            flat_obs.field = fields
-            flat_obs.exptime = exptimes
-            flat_obs.min_nexp = len(fields)
-            flat_obs.exp_set_size = len(fields)
+            # Get the FITS headers with a common start time
+            start_time = utils.current_time()
+            fits_headers = self.get_standard_headers(observation=observation)
+            fits_headers['start_time'] = utils.flatten_time(start_time)
+
+            # Take the flat field observations (blocking)
+            current_exptimes = {c: exptimes[c][-1] for c in cameras.keys() if not finished[c]}
+            camera_events = self._take_flat_observation(current_exptimes, observation,
+                                                        fits_headers=fits_headers)
```

Comment:
Need to check safety inside the blocking call.

```suggestion
            camera_events = self._take_flat_observation(current_exptimes, observation, safety_func
                                                        fits_headers=fits_headers)
```

Reply:
At the moment (with your suggestion about checking safety while taking darks), safety is checked before entering `_take_flat_observation`. I think that is sufficient, unless we want to interrupt exposures?